File: bigtangle-typescript/src/net/bigtangle/core/TokenInfo.ts
import { DataClass } from './DataClass';
import { Token } from './Token';
import { MultiSignAddress } from './MultiSignAddress';
import { Json } from '../utils/Json';

export class TokenInfo extends DataClass {
    private token: Token | null = null;
    private multiSignAddresses: MultiSignAddress[] = [];

    constructor() {
        super();
        this.multiSignAddresses = [];
    }

    public toByteArray(): Uint8Array {
        try {
            const jsonStr = Json.jsonmapper().writeValueAsString(this);
            return new TextEncoder().encode(jsonStr);
        } catch (e: any) {
            throw new Error(e);
        } 
    }

    public parse(buf: Uint8Array): TokenInfo {
        const jsonStr = new TextDecoder('utf-8').decode(buf);
        return Json.jsonmapper().readValue(jsonStr, TokenInfo);
    }

    public parseChecked(buf: Uint8Array): TokenInfo {
        const jsonStr = new TextDecoder('utf-8').decode(buf);
        try {
            return Json.jsonmapper().readValue(jsonStr, TokenInfo);
        } catch (e: any) {
            throw new Error(e);
        }
    }

    public getToken(): Token | null {
        return this.token;
    }

    public setToken(token: Token | null): void {
        this.token = token;
    }

    public getMultiSignAddresses(): MultiSignAddress[] {
        return this.multiSignAddresses;
    }

    public setMultiSignAddresses(multiSignAddresses: MultiSignAddress[]): void {
        this.multiSignAddresses = multiSignAddresses;
    }

    public toString(): string {
        return `TokenInfo [tokens=${this.token}, multiSignAddresses=${this.multiSignAddresses}]`;
    }
}

File: bigtangle-core/src/main/java/net/bigtangle/core/TokenInfo.java
/*******************************************************************************
 *  Copyright   2018  Inasset GmbH. 
 *  
 *******************************************************************************/

package net.bigtangle.core;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class TokenInfo extends DataClass implements java.io.Serializable {

    private static final long serialVersionUID = 1554582498768357964L;

    private Token token;
    private List<MultiSignAddress> multiSignAddresses;

    public byte[] toByteArray() {
        try {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            DataOutputStream dos = new DataOutputStream(baos);
            if (token != null) {
                byte[] tBytes = token.toByteArray();
                dos.writeInt(tBytes.length);
                dos.write(tBytes);
            } else {
                dos.writeInt(0);
            }
            if (multiSignAddresses != null) {
                dos.writeInt(multiSignAddresses.size());
                for (MultiSignAddress msa : multiSignAddresses) {
                    byte[] mBytes = msa.toByteArray();
                    dos.writeInt(mBytes.length);
                    dos.write(mBytes);
                }
            } else {
                dos.writeInt(0);
            }
            dos.close();
            return baos.toByteArray();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public TokenInfo parse(byte[] buf) throws IOException {
        DataInputStream dis = new DataInputStream(new ByteArrayInputStream(buf));
        int tLen = dis.readInt();
        if (tLen > 0) {
            byte[] tBytes = new byte[tLen];
            dis.readFully(tBytes);
            this.token = new Token().parse(tBytes);
        }
        int mSize = dis.readInt();
        this.multiSignAddresses = new ArrayList<>();
        for (int i = 0; i < mSize; i++) {
            int mLen = dis.readInt();
            byte[] mBytes = new byte[mLen];
            dis.readFully(mBytes);
            this.multiSignAddresses.add(MultiSignAddress.parse(mBytes));
        }
        dis.close();
        return this;
    }

    public TokenInfo parseChecked(byte[] buf) {
        try {
            return parse(buf);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public Token getToken() {
        return token;
    }

    public void setToken(Token tokens) {
        this.token = tokens;
    }

    public List<MultiSignAddress> getMultiSignAddresses() {
        return multiSignAddresses;
    }

    public void setMultiSignAddresses(List<MultiSignAddress> multiSignAddresses) {
        this.multiSignAddresses = multiSignAddresses;
    }

    public TokenInfo() {
        this.multiSignAddresses = new ArrayList<>();
    }

    @Override
    public String toString() {
        return "TokenInfo [tokens=" + token + ", multiSignAddresses=" + multiSignAddresses + "]";
    }
}
